pkg/collector/corechecks/snmp/discovery/discovery.go

```diff
@@ -0,0 +1,294 @@
+	log.Debugf("subnet %s: Start SNMP worker %d", d.config.Network, w)
+	for {
+		select {
+		case <-d.stop:
+			log.Debugf("subnet %s: Stop SNMP worker %d", d.config.Network, w)
+			return
+		case job := <-jobs:
+			log.Debugf("subnet %s: Handling IP %s", d.config.Network, job.currentIP.String())
+			err := d.checkDevice(job)
+			if err != nil {
+				log.Errorf(err.Error())
+			}
+		}
+	}
+}
+
+func (d *Discovery) checkDevices() {
+	ipAddr, ipNet, err := net.ParseCIDR(d.config.Network)
+	if err != nil {
+		log.Errorf("subnet %s: Couldn't parse SNMP network: %s", d.config.Network, err)
+		return
+	}
+
+	startingIP := ipAddr.Mask(ipNet.Mask)
+
+	configHash := d.config.DiscoveryDigest(d.config.Network)
+	cacheKey := fmt.Sprintf("%s:%s", cacheKeyPrefix, configHash)
+
+	subnet := snmpSubnet{
+		config:         d.config,
+		startingIP:     startingIP,
+		network:        *ipNet,
+		cacheKey:       cacheKey,
+		devices:        map[string]string{},
+		deviceFailures: map[string]int{},
+	}
+
+	d.loadCache(&subnet)
+
+	jobs := make(chan snmpJob)
+	for w := 0; w < d.config.DiscoveryWorkers; w++ {
+		go d.runWorker(w, jobs)
+	}
+
+	discoveryTicker := time.NewTicker(time.Duration(d.config.DiscoveryInterval) * time.Second)
+
+	for {
+		log.Debugf("subnet %s: Run discovery", d.config.Network)
+		startingIP := make(net.IP, len(subnet.startingIP))
+		copy(startingIP, subnet.startingIP)
+		for currentIP := startingIP; subnet.network.Contains(currentIP); incrementIP(currentIP) {
+
+			if ignored := subnet.config.IsIPIgnored(currentIP); ignored {
+				continue
+			}
+
+			jobIP := make(net.IP, len(currentIP))
+			copy(jobIP, currentIP)
+			job := snmpJob{
+				subnet:    &subnet,
+				currentIP: jobIP,
+			}
+			jobs <- job
+
+			select {
+			case <-d.stop:
+				log.Debugf("subnet %s: Stop scheduling devices", d.config.Network)
+				return
+			default:
+			}
+		}
+
+		select {
+		case <-d.stop:
+			log.Debugf("subnet %s: Stop scheduling devices", d.config.Network)
+			return
+		case <-discoveryTicker.C:
+		}
+	}
+}
+
+func (d *Discovery) checkDevice(job snmpJob) error {
+	deviceIP := job.currentIP.String()
+	config := *job.subnet.config // shallow copy
+	config.IPAddress = deviceIP
+	sess, err := session.NewSession(&config)
+	if err != nil {
+		return fmt.Errorf("error configure session for ip %s: %v", deviceIP, err)
+	}
+	entityID := job.subnet.config.DiscoveryDigest(deviceIP)
+	if err := sess.Connect(); err != nil {
+		log.Debugf("subnet %s: SNMP connect to %s error: %v", d.config.Network, deviceIP, err)
+		d.deleteDevice(entityID, job.subnet)
+	} else {
+		defer sess.Close()
+
+		oids := []string{sysObjectIDOid}
+		// Since `params<GoSNMP>.ContextEngineID` is empty
+		// `params.Get` might lead to multiple SNMP GET calls when using SNMP v3
+		value, err := sess.Get(oids)
+		if err != nil {
+			log.Debugf("subnet %s: SNMP get to %s error: %v", d.config.Network, deviceIP, err)
+			d.deleteDevice(entityID, job.subnet)
```

Comment:
is this deleting the device from the cache?

Reply:
Could help adding a comment saying `removing so that we don't try to discover it again`

Reply:
Added a comment here 0295e8e7a

> is this deleting the device from the cache?

It's deleting from in-mem and cache

> Could help adding a comment saying `removing so that we don't try to discover it again`

That's not the cache, this is only removing the device from the discovered devices list but I can be rediscovered again.